ARCs/arc-0024.md

```diff
@@ -0,0 +1,485 @@
+
+JSON formats, as above, are both human and machine readable.
+
+The specification MUST always be both human and machine readable.
+
+Extensions to this specification MUST NOT ever include machine-only or binary formats.
+
+### **3. Extensibility**
+
+It is difficult to know what the Application Specification will be expected
+to contain in the future. One thing that is reasonably certain though
+is that everyone would benefit if the information about the Application
+was not scattered across different places in different formats.
+
+It may not always be desirable to add information directly into the Application
+Specification file either, in which case the file can be extended using
+pointers to yet further resources, such as IPFS ids, URLs and so on.
+
+The need for extensibility means that the modified _Contract_ element in 1 above
+is insufficient as an Application Specification alone. This ARC proposes an
+envelope as follows, again following a similar descriptive style as in ARC-4:
+
+```typescript
+interface ApplicationSpecification {
+
+  /** The version  of the specification format **/
+  specificationVersion: string;  
+
+  /** The contract implemented by this Application **/
+  contract: Contract;
+
+  /** ARC extension 1**/
+  extensionPoint1: SomeFutureExtension1; 
+
+    ...
+
+  /** ARC extension N**/
+  extensionPoint2: SomeFutureExtensionN; 
+
+}
+```
+
+This ARC proposes, therefore, that Application Specification
+formats are versioned and subsequent ARCs increment the version number.
+
+### **4. Discovery**
+
+Consumers of an Application need to be able to find the Application Specification
+for that version of it.
+
+There are many potential 'user stories' around searching for different Applications.
+We could speculate on various needs, such as:
+
+* Search by a well known Interface name that the Application implements,
+* Search by a tag,
+* Search by a foreign language tag,
+* Search if the Application is audited and by whom,
+* By transaction fees,
+* By a community 'star' rating,
+* Etc.
+
+It is difficult to know where to draw the line between what should be community
+add-on, and what should be offered by the Algorand platform.
+
+In any case, there are some important security considerations:
+
+* The consumer needs to know that the Application author is the publisher
+of the Application Specification.
+* Whichever way the consumer obtains the Application ID on the
+Algorand blockchain, they need a mechanism to find the corresponding
+Application Specification.
+* When peripheral directories and explorers of Application
+metadata emerge, users may be exposed to phishing-type attacks where
+they are misdirected to fake application clones (eg: T1nyman, spelt
+with a 'one').
+
+At a minimum then some kind of veriable link between Author, Application and
+Specification needs to exist.
+
+This ARC proposes the following:
+
+1. Store the SHA-512/256 hash of the current Application Specification file
```

Comment:
We might need to synchronize with PR for [ARC-23](https://github.com/algorandfoundation/ARCs/pull/80).

Reply:
Yes, please.  This is evidence for having a general approach in ARC23 - one that allows specifying several pieces of metadata.  I think the `bytecblock` approach would would well, perhaps with a first key that specifies "ARC24", and then pairs of bytes that describe what is being annotated.

`bytecblock "ARC24" "source" "4HSF68GH" "ARC30spec" "63GF53KU" ...`